### PR TITLE
feat(nodejs): add v9 pnpm lock file support

### LINKED ISSUE #6509: feat(pnpm): add support of Lockfile v9
## Description
`pnpm` released lockfile v9 (see https://github.com/pnpm/pnpm/pull/7666, https://github.com/pnpm/pnpm/pull/7861).
We need to add support for this.

### Discussed in https://github.com/aquasecurity/trivy/discussions/6503

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
add v9 pnpm lock file support.

- `Dev` field used for v9

## Related issues
- Close #6509 

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/docs/coverage/language/nodejs.md [hunk 1]
```diff
@@ -55,8 +55,8 @@ By default, Trivy doesn't report development dependencies. Use the `--include-de
 ### pnpm
 Trivy parses `pnpm-lock.yaml`, then finds production dependencies and builds a [tree][dependency-graph] of dependencies with vulnerabilities.
 
-!!! note
-    Trivy currently only supports Lockfile [v6][pnpm-lockfile-v6] or earlier.
+#### lock file v9 version
+Trivy supports `Dev` field for `pnpm-lock.yaml` v9 or later. Use the `--include-dev-deps` flag to include the developer's dependencies in the result.
```
Comment (DmitriyLewen) [inline docs/docs/coverage/language/nodejs.md:59]: I have not seen any requests to use `Dev` field for version `6.0` or earlier lock files.
But if users want, we can add this in another PR.

File: pkg/dependency/parser/nodejs/pnpm/parse.go [hunk 2]
```diff
@@ -2,10 +2,12 @@ package pnpm
 
 import (
 	"fmt"
+	"sort"
 	"strconv"
 	"strings"
 
 	"github.com/samber/lo"
+	"golang.org/x/exp/maps"
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/pnpm/parse.go:10]:
```suggestion
	"maps"
```

File: pkg/dependency/parser/nodejs/pnpm/parse.go [hunk 3]
```diff
@@ -127,55 +252,102 @@ func (p *Parser) parseLockfileVersion(lockFile LockFile) float64 {
...
 }
 
-func (p *Parser) parseDepPath(depPath, versionSep string) (string, string) {
-	// Skip registry
-	// e.g.
-	//    - "registry.npmjs.org/lodash/4.17.10" => "lodash/4.17.10"
-	//    - "registry.npmjs.org/@babel/generator/7.21.9" => "@babel/generator/7.21.9"
-	//    - "/lodash/4.17.10" => "lodash/4.17.10"
-	_, depPath, _ = strings.Cut(depPath, "/")
+// trimRegistry trims registry (or `/` prefix) for depPath.
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/pnpm/parse.go:274]: PURL has `vcs_url`. I think we should keep this information if it is not `registry.npmjs.org`.
https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm

File: pkg/dependency/parser/nodejs/pnpm/parse.go [hunk 4]
```diff
@@ -107,6 +141,97 @@ func (p *Parser) parse(lockVer float64, lockFile LockFile) ([]ftypes.Package, []
...
+		var dependsOn []string
+		for depName, depVer := range lo.Assign(snapshot.OptionalDependencies, snapshot.Dependencies) {
+			depVer = p.trimPeerDeps(depVer, lockVer) // pnpm has already separated dep name. therefore, we only need to separate peer deps.
+			depVer = p.parseVersion(depPath, depVer, lockVer)
+			id := packageID(depName, depVer)
+			if _, ok := lockFile.Packages[id]; ok {
+				dependsOn = append(dependsOn, id)
+			}
+		}
+		if dependsOn != nil {
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/pnpm/parse.go:163]: nit: https://go.dev/play/p/DtIqxxy-mLq
```suggestion
		if len(dependsOn) != 0 {
```

Comment (knqyf263) [reply pkg/dependency/parser/nodejs/pnpm/parse.go:163]: BTW: I want golangci-lint to enforce it.

File: pkg/dependency/parser/nodejs/pnpm/parse.go [hunk 5]
```diff
@@ -107,6 +141,97 @@ func (p *Parser) parse(lockVer float64, lockFile LockFile) ([]ftypes.Package, []
...
+		if !pkg.Dev {
+			p.markRootPkgs(pkg.ID, resolvedPkgs, resolvedDeps)
+		}
+	}
+
+	return maps.Values(resolvedPkgs), maps.Values(resolvedDeps)
+}
+
+// markRootPkgs sets `Dev` to false for non dev dependency.
+func (p *Parser) markRootPkgs(id string, libs map[string]ftypes.Package, deps map[string]ftypes.Dependency) {
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/pnpm/parse.go:219]: nit
```suggestion
func (p *Parser) markRootPkgs(id string, pkgs map[string]ftypes.Package, deps map[string]ftypes.Dependency) {
```

File: pkg/dependency/parser/nodejs/pnpm/parse.go [hunk 6]
```diff
@@ -107,6 +141,97 @@ func (p *Parser) parse(lockVer float64, lockFile LockFile) ([]ftypes.Package, []
...
+	lib, ok := libs[id]
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/pnpm/parse.go:220]:
```suggestion
	pkg, ok := pkgs[id]
```

**Review by knqyf263 — COMMENTED** [review]
Looks almost good

File: pkg/dependency/parser/nodejs/pnpm/parse.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go:10]: This is weird, but `maps` package doesn't have `Values` and `Keys` functions:
https://pkg.go.dev/maps

File: pkg/dependency/parser/nodejs/pnpm/parse.go (hunk 5 above)
Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: Thanks! fixed in https://github.com/aquasecurity/trivy/pull/6617/commits/1ebe152d278656cbab059a9678d7d359d65ddd9c

File: pkg/dependency/parser/nodejs/pnpm/parse.go (hunk 6 above)
Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: fixed in https://github.com/aquasecurity/trivy/pull/6617/commits/1ebe152d278656cbab059a9678d7d359d65ddd9c

File: pkg/dependency/parser/nodejs/pnpm/parse.go (hunk 4 above)
Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: fixed in https://github.com/aquasecurity/trivy/pull/6617/commits/535b1e22b44c990578f3f04ea88b235abf0b28be

File: pkg/dependency/parser/nodejs/pnpm/parse.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: updated in https://github.com/aquasecurity/trivy/pull/6617/commits/1dc202a7556f11d0835aa56bec43d11a84954a2a

Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: I am not sure if using registry URL in package name is correct.
I suggest keeping depPath in [ExternalReferences](https://github.com/aquasecurity/trivy/blob/bbaf5952bc8059c537401bd4364e019eecd16c9a/pkg/fanal/types/package.go#L173)
and check `ExternalReferences` in purl for npm:
https://github.com/aquasecurity/trivy/blob/bbaf5952bc8059c537401bd4364e019eecd16c9a/pkg/purl/purl.go#L425-L429

e.g.
`private.npmjs.org/@babel/runtime@7.18.3` =>
```Go
ID:           "@babel/runtime@7.18.3",
Name:         "@babel/runtime",
Version:      "7.18.3",
Relationship: ftypes.RelationshipIndirect,
ExternalReferences: []ftypes.ExternalRef{
	{
		Type: ftypes.RefVCS,
		URL:  "private.npmjs.org/@babel/runtime@7.18.3",
	},
},
```

@knqyf263 wdyt?

Comment (knqyf263) [reply pkg/dependency/parser/nodejs/pnpm/parse.go]: Yes, I meant to suggest using `ExternalReferences`, lack of explanation.

Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/pnpm/parse.go:277]: I update this logic - https://github.com/aquasecurity/trivy/pull/6617/commits/2d50522c0d18104c8cb5b4c628ed3b75cc326352
take a look, when you have time, please

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**